internal/consuming/kafka_test.go

```diff
@@ -60,6 +60,36 @@ func produceTestMessage(topic string, message []byte) error {
 	return nil
 }
 
+func produceTestMessageToPartition(topic string, message []byte, partition int32) error {
+	partitioner := kgo.BasicConsistentPartitioner(func(topic string) func(*kgo.Record, int) int {
+		return func(r *kgo.Record, n int) int {
+			return int(r.Partition)
+		}
+	})
```

Comment:
I think this is just the `kgo.ManualPartitioner()`?

Reply:
Many thanks @twmb  - fixing in https://github.com/centrifugal/centrifugo/pull/823

What do you think about the approach here BTW? We originally implemented consumer based on the [example](https://github.com/twmb/franz-go/blob/master/examples/goroutine_per_partition_consuming/autocommit_marks/main.go) in `franz-go` repo, but without channel buffer for the partition consumer (i.e. `recs` channel). This quickly resulted into poll blocked when partition processing was long, and consumer excluded from group by Kafka due to poll timeout. `franz-go` uses buffered channel for a partition consumer, but it seems it has the same issue - just with less probability, as soon as channel buffer (5 in the example) is full - poll will block. I believe the approach here fully solves these concerns. WDYT?

Reply:
I think it works. It may cause more bandwidth to be used, but there isn't a way to really avoid that.

There may be data that is already buffered, or a fetch already in progress, for partitions you are pausing. The data is just dropped when being polled. Once you resume the fetch partition, the partition is re-fetched at the prior offset.

It may be worth having a buffer of 1 or 5 on the channel, to allow for the chance that processing finishes before you _need_ to pause.

Another thing to do may be some select/default case before the existing select.

```
select {
<-done:
<-quit:
processor <- records:
default:
}

cl.Pause...

select {
<-done:
<-quit:
processor <- records:
}
```

The first select default would allow you to avoid pausing if things are keeping up, then you only pause if needed when things are backed up.

Reply:
Thanks a lot, yeah - looks like using a buffer makes sense, addressing in https://github.com/centrifugal/centrifugo/pull/829